sklearn/manifold/isomap.py

```diff
@@ -96,11 +149,51 @@ def __init__(self, n_neighbors=5, n_components=2, eigen_solver='auto',
         self.neighbors_algorithm = neighbors_algorithm
         self.nbrs_ = NearestNeighbors(n_neighbors=n_neighbors,
                                       algorithm=neighbors_algorithm)
+        self.n_landmarks = n_landmarks
+        self.landmarks_ = landmarks
+        self.random_state = random_state
+
+    def _compute_landmarks(self, X):
+        """Computes the landmarks in X to use.
+
+        Parameters
+        ----------
+        X : matrix (n_samples, n_features)
+            The data set containing the samples from where the landmarks
+            should be inferred.
+
+        Returns
+        -------
+        landmarks: array-like, shape (n_landmarks,)
+        """
+        if self.landmarks_ is None:
+            n_samples = X.shape[0]
+            n_landmarks = self.n_landmarks
+
+            if n_landmarks is not None:
+                if n_landmarks == 'auto':
+                    # To guarantee some stability, the number of landmarks
+                    # should be strictly superior to the number of dimensions
+                    # of the reduced data set added by some safety margin.
+                    n_landmarks = (self.n_components + 1 +
+                                   min(self.n_components // 2, 100))
+
+                    # Limit number of landmarks by the number of samples.
+                    n_landmarks = min(n_landmarks, n_samples)
+
+                self.landmarks_ = np.arange(n_samples)
+                check_random_state(self.random_state).shuffle(self.landmarks_)
+                self.landmarks_ = self.landmarks_[:n_landmarks]
+
+        return self.landmarks_
 
     def _fit_transform(self, X):
         X = check_array(X)
+        landmarks = self._compute_landmarks(X)
+
         self.nbrs_.fit(X)
-        self.training_data_ = self.nbrs_._fit_X
+        self.training_data_ = self.nbrs_._fit_X if landmarks is None else self.nbrs_._fit_X[landmarks]
+
```

Comment:
Yes – I understand that you'll have to use this array differently, but to me it makes more sense for ``training_data`` to store the training data, rather than the landmarks.